### PR TITLE
feat(examples) Add simulation tensorflow example on tabular datasets

### PR DESCRIPTION
**Pull request by damjimenezgu**
### Description

Include an end-to-end simulation example of running a Federated Learning with Flower using a tabular dataset for training a Multi-layer Perceptron (MLP) or Deep Neural Network (DNN) classifier in TensorFlow.

### Explanation

The new example will provide a quick-start on simulating FL models with tabular data using TensorFlow

### Checklist

- [x ] Implement proposed change
- [ x] Write tests
- [x ] Update [documentation](https://flower.ai/docs/writing-documentation.html)
- [x ] Make CI checks pass
- [ x] Ping maintainers on [Slack](https://flower.ai/join-slack/) (channel `#contributions`)

### CODE REVIEW

File: examples/simulation-tensorflow-tabular/README.md [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+If you don't see any errors you're good to go!
+
+#### pip
+
+Write the command below in your terminal to install the dependencies according to the configuration file requirements.txt.
+
+```shell
+pip install -r requirements.txt
+```
```
Comment (jafermarq) [inline examples/simulation-tensorflow-tabular/README.md:54]: We are gradually removing Poetry from the examples and only use a single `pyproject.toml` (i.e. an no `requirementes.txt`) to define the dependencies. See how we have done it in [examples/quickstart-pytorch](https://github.com/adap/flower/tree/main/examples/quickstart-pytorch#installing-dependencies).

File: examples/simulation-tensorflow-tabular/requirements.txt
Comment (jafermarq) [inline examples/simulation-tensorflow-tabular/requirements.txt:1]: it seems `scikit-learn` dependency is needed too.

File: examples/simulation-tensorflow-tabular/sim.py [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+train_cds = Dataset.from_pandas(train_cds)
+
+centralized_testset = {"features": Xy_test[RELEVANT_FEATURES].values.tolist(), "labels": Xy_test["Survived"].values.tolist()}
+centralized_testset = pd.DataFrame(centralized_testset)
+centralized_testset = Dataset.from_pandas(centralized_testset).to_tf_dataset(
+    columns="features", label_cols="labels", batch_size=64
+)
+
+# Set partitioner to federate data
+partitioner = DirichletPartitioner(num_partitions=NUM_CLIENTS, partition_by="labels", alpha=ALPHA)
```
Comment (jafermarq) [inline examples/simulation-tensorflow-tabular/sim.py:193]: This looks ok but I think it should be possible to do in a more compact way with `flwr-datasets`. I'm thinking something following the order: (1) 1 generate partitions, (2) each partition is split into `traint`/`val`/`test`; (3) the `test` partitions are fused to form the `centralised_testset`. What do you think? @damjimenezgu @adam-narozniak

**Review by jafermarq — COMMENTED** [review]
@damjimenezgu great to see this example! Nice to see you added also compatibility with `flower-simulation`.

I think it would be better to rename the directory to as `tensorflow-tabular`. When you create your `pyproject.toml` could you set the minimum version of `flwr` to 1.8.0 and of `flwr-datasets` to 0.1.0 -- those are the latest versions for each package.

File: examples/simulation-tensorflow-tabular/sim.py (hunk 2 above)
Comment (adam-narozniak) [reply examples/simulation-tensorflow-tabular/sim.py:193]: I don't see any point in delegating the responsibility of downloading the dataset (in this case) to load_dataset; instead, the `FederatedDataset` can be used.
The details on how the data is divided are less relevant (I'm okay with Javier's suggestion).

File: examples/simulation-tensorflow-tabular/sim.py [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+    data = pd.concat([features,data['Survived']], axis=1)
+    
+    return data
+
+# Load dataset from Hugging Face
+titanic_cds = load_dataset("julien-c/titanic-survival")
+
+# Convert to pandas to preprocess
+titanic_cds = titanic_cds['train'].to_pandas()
+titanic_cds = preprocess_data(titanic_cds)
```
Comment (adam-narozniak) [inline examples/simulation-tensorflow-tabular/sim.py:176]: That's a leak in the preprocessing dataset. Each partition should be normalized/standardized according to ONLY the LOCAL data.

File: examples/simulation-tensorflow-tabular/sim.py [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+    # Scale the features
+    scaler = StandardScaler()
+    features = pd.DataFrame(scaler.fit_transform(features), columns=RELEVANT_FEATURES)
+
+    data = pd.concat([features,data['Survived']], axis=1)
+    
+    return data
+
+# Load dataset from Hugging Face
+titanic_cds = load_dataset("julien-c/titanic-survival")
```
Comment (adam-narozniak) [inline examples/simulation-tensorflow-tabular/sim.py:172]: What is *`cds`? If an example is designed to be for the general public, there's no point in giving non-informative names.

File: examples/simulation-tensorflow-tabular/sim.py [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+    help="Ratio of GPU memory to assign to a virtual client",
+)
+
+VERBOSE = 0
+NUM_CLIENTS = 5
+RANDOM_STATE = 42
+NUM_ROUNDS = 30
+
+# Define Alpha to use for DirichletPartitioner
+ALPHA = 1000
```
Comment (adam-narozniak) [inline examples/simulation-tensorflow-tabular/sim.py:45]: Basically, there is no point in using Dirichlet with such high value unless **heavily empathized**  reasoning. That's close to iid.

File: examples/simulation-tensorflow-tabular/sim.py [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+
+# Create FedAvg strategy
+strategy = fl.server.strategy.FedAvg(
+    fraction_fit=1.0,  # Sample 100% of available clients for training
+    fraction_evaluate=0.5,  # Sample 50% of available clients for evaluation
+    min_fit_clients=2,  # Never sample less than 2 clients for training
+    min_evaluate_clients=2,  # Never sample less than 2 clients for evaluation
+    min_available_clients=int(
+        NUM_CLIENTS * 0.75
+    ),  # Wait until at least 75 clients are available
```
Comment (adam-narozniak) [inline examples/simulation-tensorflow-tabular/sim.py:204]: not 75 but int(5 * 0.75), not sure if that's intended since there's diff between the code and the comment

File: examples/simulation-tensorflow/README.md [hunk 7]
```diff
@@ -4,7 +4,7 @@ This introductory example uses the simulation capabilities of Flower to simulate
 
 ## Running the example (via Jupyter Notebook)
 
-Run the example on Google Colab: [![Open in Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/adap/flower/blob/main/examples/simulation-tensorflow/sim.ipynb)
+Run the example on Google Colab: [![Open in Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/damjimenezgu/flower/blob/simulation-tensorflow-tabular/examples/simulation-tensorflow/sim.ipynb)
```
Comment (adam-narozniak) [inline examples/simulation-tensorflow/README.md:7]: shouldn't be commited

**Review by adam-narozniak — CHANGES_REQUESTED** [review]

File: examples/simulation-tensorflow-tabular/README.md (hunk 1 above)
Comment (damjimenezgu) [reply examples/simulation-tensorflow-tabular/README.md:54]: I changed the version of flwr to 1.8.0 and of flwr-datasets to 0.1.0. Additionally, following the example, I changed the README to remove the Poetry dependence.

PS: Check the new [PR #3633](https://github.com/adap/flower/pull/3633) with all the modifications.

File: examples/simulation-tensorflow-tabular/requirements.txt
Comment (damjimenezgu) [reply examples/simulation-tensorflow-tabular/requirements.txt:1]: I added the dependency to the .toml file

PS: Check the new [PR #3633](https://github.com/adap/flower/pull/3633) with all the modifications.

File: examples/simulation-tensorflow-tabular/sim.py (hunk 2 above)
Comment (damjimenezgu) [reply examples/simulation-tensorflow-tabular/sim.py:193]: I followed your advice using the `FederatedDataset`.

PS: Check the new [PR #3633](https://github.com/adap/flower/pull/3633) with all the modifications.

File: examples/simulation-tensorflow-tabular/sim.py (hunk 3 above)
Comment (damjimenezgu) [reply examples/simulation-tensorflow-tabular/sim.py:176]: Thanks! I modified the preprocessing step to be done ONLY with the LOCAL data.

PS: Check the new [PR #3633](https://github.com/adap/flower/pull/3633) with all the modifications.

File: examples/simulation-tensorflow-tabular/sim.py (hunk 4 above)
Comment (damjimenezgu) [reply examples/simulation-tensorflow-tabular/sim.py:172]: It was removed to avoid non-informative names.

PS: Check the new [PR #3633](https://github.com/adap/flower/pull/3633) with all the modifications.

File: examples/simulation-tensorflow-tabular/sim.py (hunk 5 above)
Comment (damjimenezgu) [reply examples/simulation-tensorflow-tabular/sim.py:45]: I changed it to 0.5.

PS: Check the new [PR #3633](https://github.com/adap/flower/pull/3633) with all the modifications.

File: examples/simulation-tensorflow-tabular/sim.py (hunk 6 above)
Comment (damjimenezgu) [reply examples/simulation-tensorflow-tabular/sim.py:204]: Thanks! I changed it to 3 to be congruent.

PS: Check the new [PR #3633](https://github.com/adap/flower/pull/3633) with all the modifications.

File: examples/simulation-tensorflow/README.md (hunk 7 above)
Comment (damjimenezgu) [reply examples/simulation-tensorflow/README.md:7]: Thanks, I avoided commiting it.

PS: Check the new [PR #3633](https://github.com/adap/flower/pull/3633) with all the modifications.
